### PR TITLE
Fix exponentiation for `NaNMath.pow`

### PR DESCRIPTION
**Pull request by jClugstor**
Currently an example like this:
```julia
using ForwardDiff
using NaNMath

function new_pow(x)
    NaNMath.pow(x[1],x[2])
end

ForwardDiff.gradient(new_pow, [-1.0, 2.0])
```
errors. I'd like to make it so that it returns a `NaN` instead. 

Fixes https://github.com/JuliaDiff/ForwardDiff.jl/issues/716

### CODE REVIEW

**Comment by mcabbott** [comment]
Can you add a few cases to the tests? Ideally NaN/not/negative through both paths.

File: src/dual.jl [hunk 1]
```diff
@@ -552,43 +552,73 @@ end
 # exponentiation #
 #----------------#
 
-for f in (:(Base.:^), :(NaNMath.pow))
```
Comment (devmotion) [inline src/dual.jl:555]: Could we avoid the code duplication and just change this to
```julia
for (f, invf) in ((:(Base.:^), :(Base.log)), (:(NaNMath.pow), :(NaNMath.log)))
```
and use `$invf` instead of `log` below?

Comment (jClugstor) [reply src/dual.jl:555]: Yes

File: src/dual.jl [hunk 2]
```diff
@@ -583,13 +583,16 @@ for f in (:(Base.:^), :(NaNMath.pow))
...
+                deriv = (iszero(x) && v > 0) ? zero(expv) : expv*($log)(x)
                 return Dual{Ty}(expv, deriv * partials(y))
             end
         )
     end
 end
 
+
+
+
```
Comment (devmotion) [inline src/dual.jl:595]:
```suggestion
```

File: test/GradientTest.jl [hunk 3]
```diff
@@ -200,6 +201,18 @@ end
...
+    @test isnan(ForwardDiff.gradient(x -> NaNMath.pow(x[1],x[1]), [NaN, 1.0])[1])
+    @test ForwardDiff.gradient(x -> NaNMath.pow(x[1], x[2]), [1.0, 1.0]) == [1.0, 0.0]
+    @test isnan(ForwardDiff.gradient((x) -> NaNMath.pow(x[1], x[2]), [-1.0, 0.5])[1])
+
+    @test isnan(ForwardDiff.gradient(x -> x[1]^x[2], [NaN, 1.0])[1])
+    @test ForwardDiff.gradient(x -> x[1]^x[2], [1.0, 1.0]) == [1.0, 0.0]
+    @test_throws DomainError ForwardDiff.gradient(x -> x[1]^x[2], [-1.0, 0.5])
+
+end
+
```
Comment (devmotion) [inline test/GradientTest.jl:214]:
```suggestion
end
```

**Review by devmotion — APPROVED** [review]
